docs/metrics-data-model.md

```diff
@@ -0,0 +1,177 @@
+}
+```
+
+#### TagMap
+
+```go
+type TagMap struct {
+    m sync.Map // potentially something better
+}
+```
+
+It creates a finite set of tags storing the pointer of Tags by its hash. In this way, it should reduce the used memory and the GC operations.
+
+#### Open questions:
+
+* Is the same structure by TagSet required? It could be useful in the case we need to compare a pair of TagSet, where we can just execute single access by hash instead of N comparisons.
+* Should it be possible for any reason to fetch raw data points (not aggregated)? Like `samples := tsdb.GetSamplesBySeries(seriesID, timeStart, timeEnd)`? 
+
+## Collect
+
+#### Open questions:
+
+* Should we change something in `metrics.PushIfNotDone()`?
+* Do we require to identify the time series before invoking `metrics.PushIfNotDone`? So, does the sample require to reference the time series?
+
+## metrics.Ingester
+
+TBD
+
+## Outputs
+
+Outputs should interact with the `tsdb` for fetching a time series snapshot. `series := tsdb.GetSeries()`.
+
+#### Open questions:
+
+* Should we query by timeframe so we can filter out time series without samples and avoid pushing the time series without a changed value?
+
+#### Aggregation
+
+Using time series from the outputs as described is already a form of aggregation. 
+
+#### Open questions:
+
+* Do we require high accurate timeframe for aggregation? Currently, the aggregation window is defined by the `PeriodicFlusher` but in the high concurrent world like `k6` it could be not exactly the expected timeframe (e.g. instead to be now-1s <= t < now).
+
+#### Concurrency
+
+In the case, that the write endpoint for flushing data is slow and it returns the response in a greater interval than the interval set for the `PeriodicFlusher` then we could push a not optimal aggregated window where the aggregation is based on a wider range.
+
+Is this an issue? If yes:
+
+* In the case the Prometheus instance slows down in a limited range then we could mitigate it by adding concurrent flush. It requires adding a label per goroutine (e.g. a Worker ID ?) so we can avoid hitting out-of-order issues. 
+
+#### Prometheus
+
+##### Trend
+
+[:bulb:] We could potentially mitigate the sparse histogram by making the histogram configurable by adding two options:
+
+* UseTrendAsHistogram: where it says to the output to use the histogram for Trend.
+* HistogramBuckets: it would require a map between Trend metrics' name and the expected histogram.
+
+In the case, we couldn't directly implement the right dynamic histogram, this would help advanced users that know how the SUT behaves to use histogram instead of the 6 gauges workaround.
+
+##### End of series markers
+
+Is the marker required and/or useful for the output?
+
+#### Thresholds
+
+Thresholds should act as an output, it should work on top of a time series data layer. They get the latest value for the time series and check if the value triggers the threshold based on the defined condition.
```

Comment:
This is how it already is done after my latest refactoring, see how the `MetricsEngine` receives metric samples by just pretending to be another normal `Output`:
https://github.com/grafana/k6/blob/b725f03be5873b9b0d933b94a83924ff7c1f15e4/metrics/engine/ingester.go#L14-L16
https://github.com/grafana/k6/blob/b725f03be5873b9b0d933b94a83924ff7c1f15e4/core/engine.go#L98-L103

But it can't directly work on the aggregated data from time series since a single threshold spans multiple time series.